File: ai-python/models/detected_change.py
from datetime import datetime
from sqlalchemy import Column, Integer, String, Float, DateTime, Text, create_engine
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import sessionmaker

Base = declarative_base()


class DetectedChange(Base):
    __tablename__ = "detected_changes"

    id = Column(Integer, primary_key=True)
    page_id = Column(Integer, nullable=False, index=True)
    page_type = Column(String(20), nullable=True)  # pricing, features
    
    # Price changes
    old_price = Column(String(50), nullable=True)
    new_price = Column(String(50), nullable=True)
    change_percent = Column(Float, nullable=True)
    
    # Availability changes
    old_availability = Column(String(50), nullable=True)
    new_availability = Column(String(50), nullable=True)
    
    # Feature changes
    old_features = Column(Text, nullable=True)  # JSON array of features
    new_features = Column(Text, nullable=True)
    features_added = Column(Text, nullable=True)  # JSON
    features_removed = Column(Text, nullable=True)  # JSON
    
    # Messaging changes
    old_text = Column(Text, nullable=True)
    new_text = Column(Text, nullable=True)
    
    # Change type: price_increase, price_decrease, availability_change, 
    # feature_added, feature_removed, messaging_change, content_change
    change_type = Column(String(50), nullable=False)
    
    # Hash for quick comparison
    old_hash = Column(String(64), nullable=True)
    new_hash = Column(String(64), nullable=True)
    
    detected_at = Column(DateTime, default=datetime.utcnow, nullable=False)
    raw_data = Column(Text, nullable=True)

    def __repr__(self):
        return f"<DetectedChange(id={self.id}, page_id={self.page_id}, change_type={self.change_type})>"

File: api-go/models/detected_change.go
package models

import "time"

// DetectedChange mirrors the detected_changes table (written by Python AI engine)
type DetectedChange struct {
	ID              uint      `gorm:"primaryKey" json:"id"`
	PageID          int       `gorm:"column:page_id;not null;index" json:"page_id"`
	PageType        string    `gorm:"column:page_type;type:varchar(20)" json:"page_type"`
	ChangeType      string    `gorm:"column:change_type;type:varchar(50);not null" json:"change_type"`
	OldPrice        string    `gorm:"column:old_price;type:varchar(50)" json:"old_price"`
	NewPrice        string    `gorm:"column:new_price;type:varchar(50)" json:"new_price"`
	ChangePercent   float64   `gorm:"column:change_percent" json:"change_percent"`
	OldAvailability string    `gorm:"column:old_availability;type:varchar(50)" json:"old_availability"`
	NewAvailability string    `gorm:"column:new_availability;type:varchar(50)" json:"new_availability"`
	OldFeatures     string    `gorm:"column:old_features;type:text" json:"old_features"`
	NewFeatures     string    `gorm:"column:new_features;type:text" json:"new_features"`
	FeaturesAdded   string    `gorm:"column:features_added;type:text" json:"features_added"`
	FeaturesRemoved string    `gorm:"column:features_removed;type:text" json:"features_removed"`
	OldText         string    `gorm:"column:old_text;type:text" json:"old_text"`
	NewText         string    `gorm:"column:new_text;type:text" json:"new_text"`
	OldHash         string    `gorm:"column:old_hash;type:varchar(64)" json:"old_hash"`
	NewHash         string    `gorm:"column:new_hash;type:varchar(64)" json:"new_hash"`
	DetectedAt      time.Time `gorm:"column:detected_at;not null" json:"detected_at"`
	RawData         string    `gorm:"column:raw_data;type:text" json:"raw_data"`
}

func (DetectedChange) TableName() string {
	return "detected_changes"
}
